### PR TITLE
Fix memory leaks in PeerConnection

### LINKED ISSUE #425: Memory leaks in PeerConnection
There are memory leaks in the following methods of PeerConnection:
  * CreateOffer
  * CreateAnswer
  * AddIceCandidate
  * GetStats

They can be found by executing the unit tests in `valgrind`:

In `create-offer.js`:
```text
=7791== 244 (64 direct, 180 indirect) bytes in 1 blocks are definitely lost in loss record 497 of 609
==7791==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==7791==    by 0x9F87E82: webrtc::WebRtcSessionDescriptionFactory::InternalCreateOffer(webrtc::CreateSessionDescriptionRequest) (in /tmp/node-webrtc/build/Debug/wrtc.node)
==7791==    by 0x9F8663C: webrtc::WebRtcSessionDescriptionFactory::SetCertificate(rtc::scoped_refptr<rtc::RTCCertificate> const&) (in /tmp/node-webrtc/build/Debug/wrtc.node)
==7791==    by 0x9F85B73: webrtc::WebRtcCertificateGeneratorCallback::OnSuccess(rtc::scoped_refptr<rtc::RTCCertificate> const&) (in /tmp/node-webrtc/build/Debug/wrtc.node)
==7791==    by 0x9CBB1AF: rtc::(anonymous namespace)::RTCCertificateGenerationTask::OnMessage(rtc::Message*) (in /tmp/node-webrtc/build/Debug/wrtc.node)
==7791==    by 0x9CC05D9: rtc::MessageQueue::Dispatch(rtc::Message*) (in /tmp/node-webrtc/build/Debug/wrtc.node)
==7791==    by 0x9B83E3B: rtc::Thread::Run() (in /tmp/node-webrtc/build/Debug/wrtc.node)
==7791==    by 0x9B83B83: rtc::Thread::PreRun(void*) (in /tmp/node-webrtc/build/Debug/wrtc.node)
==7791==    by 0x5B936DA: start_thread (pthread_create.c:463)
==7791==    by 0x5ECC88E: clone (clone.S:95)
```

In `connect.js`:
```text
==17370== 208 (64 direct, 144 indirect) bytes in 2 blocks are definitely lost in loss record 781 of 1,079
==17370==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==17370==    by 0x9B22D51: node_webrtc::PeerConnection::GetStats(Nan::FunctionCallbackInfo<v8::Value> const&) (peerconnection.cc:522)
==17370==    by 0x9AEADBC: Nan::imp::FunctionCallbackWrapper(v8::FunctionCallbackInfo<v8::Value> const&) (nan_callbacks_12_inl.h:174)
==17370==    by 0xA5FD82: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) (in /home/scote/.nvm/versions/node/v8.11.2/bin/node)
==17370==    by 0xAD6F2B: v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) (in /home/scote/.nvm/versions/node/v8.11.2/bin/node)
```

```text
==17370== 1,466 (768 direct, 698 indirect) bytes in 3 blocks are definitely lost in loss record 962 of 1,079
==17370==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==17370==    by 0x9BC3821: webrtc::CreateIceCandidate(std::string const&, int, std::string const&, webrtc::SdpParseError*) (in /tmp/node-webrtc/build/Debug/wrtc.node)
==17370==    by 0x9A9123F: CreateIceCandidateInterface(std::string const&, std::string const&, int, node_webrtc::Maybe<std::string> const&) (webrtc.cc:502)
==17370==    by 0x9ABF8D3: std::_Function_handler<node_webrtc::Validation<webrtc::IceCandidateInterface*> (std::string const&, std::string const&, int, node_webrtc::Maybe<std::string> const&), node_webrtc::Validation<webrtc::IceCandidateInterface*> (*)(std::string const&, std::string const&, int, node_webrtc::Maybe<std::string> const&)>::_M_invoke(std::_Any_data const&, std::string const&, std::string const&, int&&, node_webrtc::Maybe<std::string> const&) (std_function.h:302)
==17370==    by 0x9ABFAA2: std::function<node_webrtc::Validation<webrtc::IceCandidateInterface*> (std::string const&, std::string const&, int, node_webrtc::Maybe<std::string> const&)>::operator()(std::string const&, std::string const&, int, node_webrtc::Maybe<std::string> const&) const (std_function.h:706)
```

### PR DESCRIPTION
**Pull request by sebas642**
* CreateOffer
  * CreateAnswer
  * AddIceCandidate
  * GetStats

Fixes #425

### CODE REVIEW

File: src/peerconnection.cc [hunk 1]
```diff
@@ -426,6 +425,7 @@ NAN_METHOD(PeerConnection::AddIceCandidate) {
     self->Dispatch(std::move(promise));
   }
 
+  delete candidate;
```
Comment (markandrus) [inline src/peerconnection.cc:428]: Nice 👍

File: src/createsessiondescriptionobserver.cc [hunk 2]
```diff
@@ -26,6 +26,7 @@ void CreateSessionDescriptionObserver::OnSuccess(webrtc::SessionDescriptionInter
     }
     parent->Dispatch(std::move(_promise));
   }
+  delete sdp;
```
Comment (markandrus) [inline src/createsessiondescriptionobserver.cc:29]: Didn't realize this was a part of the contract! 👍I checked [the WebRTC source](https://chromium.googlesource.com/external/webrtc/stable/talk/+/master/app/webrtc/jsep.h) and can confirm:

```c++
// The implementation of the CreateSessionDescriptionObserver takes
// the ownership of the \|desc\|.
virtual void OnSuccess(SessionDescriptionInterface* desc) = 0;
```

**Review by markandrus — CHANGES_REQUESTED** [review]
Thanks a lot for these changes! The build is failing due to this:

```
You must run make format before submitting a pull request
diff --git a/src/peerconnection.cc b/src/peerconnection.cc
index 55214a7..ac7b596 100644
--- a/src/peerconnection.cc
+++ b/src/peerconnection.cc
@@ -522,7 +522,7 @@ NAN_METHOD(PeerConnection::GetStats) {
   if (self->_jinglePeerConnection != nullptr) {
     auto statsObserver = new rtc::RefCountedObject<StatsObserver>(self, std::move(promise));
     if (!self->_jinglePeerConnection->GetStats(statsObserver, nullptr,
-        webrtc::PeerConnectionInterface::kStatsOutputLevelStandard)) {
+            webrtc::PeerConnectionInterface::kStatsOutputLevelStandard)) {
       auto error = Nan::Error("Failed to execute getStats");
       resolver->Reject(error);
     }
```

Can you run `make format` (or edit manually) and update the PR?

**Comment by sebas642** [comment]
I have fixed the formatting. Sorry about that!

**Review by markandrus — APPROVED** [review]

**Comment by markandrus** [comment]
Thanks! 🎉

**Pull request merged**